builder/common/step_network_info.go

```diff
@@ -231,4 +184,138 @@ func (s *StepNetworkInfo) Run(ctx context.Context, state multistep.StateBag) mul
 	return multistep.ActionContinue
 }
 
+func (s *StepNetworkInfo) TryGetDefaultVPCAndSubnet(ui packersdk.Ui, ec2conn ec2iface.EC2API, state multistep.StateBag) error {
+	ui.Say(fmt.Sprintf("Setting public IP address to %t on instance without a subnet ID",
+		*s.AssociatePublicIpAddress.ToBoolPointer()))
+
+	var vpc = s.VpcId
+	if vpc == "" {
+		ui.Say("No VPC ID provided, Packer will use the default VPC")
+		vpcs, err := ec2conn.DescribeVpcs(&ec2.DescribeVpcsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("is-default"),
+					Values: []*string{aws.String("true")},
+				},
+			},
+		})
+		if err != nil {
+			return fmt.Errorf("Failed to describe VPCs: %s", err)
+		}
+
+		if len(vpcs.Vpcs) != 1 {
+			return fmt.Errorf("No default VPC found")
+		}
+		vpc = *vpcs.Vpcs[0].VpcId
+	}
+
+	var err error
+
+	ui.Say(fmt.Sprintf("Inferring subnet from the selected VPC %q", vpc))
+	params := &ec2.DescribeSubnetsInput{}
+	filters := map[string]string{
+		"vpc-id": vpc,
+		"state":  "available",
+	}
+	params.Filters, err = buildEc2Filters(filters)
+	if err != nil {
+		return fmt.Errorf("Failed to prepare subnet filters: %s", err)
+	}
+	subnetOut, err := ec2conn.DescribeSubnets(params)
+	if err != nil {
+		return fmt.Errorf("Failed to describe subnets: %s", err)
+	}
+
+	subnets := subnetOut.Subnets
+
+	// Filter by AZ with support for machine type
+	azs := getAZFromSubnets(subnets)
+	azs, err = filterAZByMachineType(azs, s.RequestedMachineType, ec2conn)
+	if err != nil {
+		return fmt.Errorf("Failed to filter subnets by AZ for machine type %q: %s",
+			s.RequestedMachineType,
+			err)
+	}
+	subnets = filterSubnetsByAZ(subnets, azs)
+	if subnets == nil {
+		return fmt.Errorf("Failed to get subnets for the filtered AZs")
+	}
+
+	subnet := mostFreeSubnet(subnets)
+
+	s.SubnetId = *subnet.SubnetId
+	s.VpcId = vpc
+
```

Comment:
It's how it's conceived now yes, though it should indeed be possible to get the AZ from the returned subnet, and we don't need to set the rest afterwards. I think it can't hurt, so I'll amend this code to set the AZ at the same time.